### ISSUE TITLE
Increased API Coverage

### ISSUE THREAD

**Issue by milseman** [apple/swift-system#16]
This is meant to be a continuously-updated roadmap for adding more API and coverage to System.

(I have just made a couple arbitrary 

### Completed

* [0.0.1](https://github.com/apple/swift-system/releases/tag/0.0.1)
  * syscalls: `read`, `write`, `lseek`, `pread`, `pwrite`, `open`, `close`
  * types: `FilePath`, `FileDescriptor`, `Errno`, `FilePermissions`
  * helpers: `closeAfter`, `writeAll`

### In Progress

* [FilePath Syntactic Operations](https://github.com/apple/swift-system/pull/14)
  * decompose a path, analyze a path, components view, mutate a path, etc

### Near Term

Anything here could be added to System in the near term. Most of the functionality can be added by following existing API design patterns, though some will need new patterns.

* sleep capabilities (e.g. `usleep`)
* FilePath semantic (i.e. file-system interacting) operations
  * `stat`, `chown`, directory iteration, etc.
  * temporary files
* `pthread` low-level interfaces
* `posix_spawn` low-level interfaces
  * signal sets
  * `execv` functionality
* Sockets
* Environment variables
  * FilePath: `~` expansion, `currentWorkingDirectory`
* OSString-like abstraction
  * Null-terminated bytes on Unix, wchar or bytes on Windows (pending investigation)
* I/O events
  * `kqueue`/`kevent` for Darwin, `epoll` for Linux, APC (or something similar) for Windows
* exit handling
  * `exit`, `atexit`, etc
* tty
  * `ioctl`, etc

### Long Term (vague "blue-sky" hopes)

* moveonly `File` type
* high level `Process` type
* `io_uring` on Linux

**Comment by allenhumphreys** [comment]
Not sure if it's on the road map, but additional `FileDescriptor` operations such as `dup` and `dup2` also support for the `FILE` APIs? I'm not 100% sure if this aligns with this packages goals, just commenting because I was recently doing some things with these.

**Comment by milseman** [comment]
@allenhumphreys you're right, I'll add those. Sorry that I didn't see your reply earlier, I sometimes miss Github notifications (still trying to figure out why).

**Comment by allenhumphreys** [comment]
Thanks @miles. For reference, [here's](https://gist.github.com/allenhumphreys/e235af773a6685f2a872d40c47770796) where I used the package. It's a great example of where the API's coverage is quite low.

Really glad to see this package coming along though!

**Comment by milseman** [comment]
Updated to reflect merged API that is yet-to-be-released, and added in references to [Standard file descriptors](https://github.com/apple/swift-system/pull/21), and [Strongly-typed process and signal wrappers](https://github.com/apple/swift-system/pull/20).

**Comment by milseman** [comment]
Updated for [dup/dup2](https://github.com/apple/swift-system/pull/28)

**Comment by GeorgeLyon** [comment]
Just curious, does `pipe` fit somewhere on this roadmap?

**Comment by milseman** [comment]
I had mentally lumped that under process types and calls, but it's not actually blocked on any design work there. We can add this anytime and might make a good starter task.

```swift
extension FileDescriptor {
  static func openPipe() throws -> (read: FileDescriptor, write: FileDescriptor)
}
```

**Comment by compnerd** [comment]
Pipes on Windows are somewhat interesting in that they behave differently.  I think that it may be a good idea to ensure that we have a proper design to accommodate that properly.

**Comment by milseman** [comment]
@compnerd, would [CreatePipe](https://docs.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-createpipe) be an appropriate equivalent? It seems sensible to have an additional overload on Windows that takes a buffer size and perhaps security policy (or at least a `Bool` for inheritability). Buffering has a default value and we can pick a default regarding inheritance for the parameter-less (or defaulted) API.

**Comment by compnerd** [comment]
`CreatePipe` can be used to create the anonymous pipe and that is better because you can control inheritability unlike with `_pipe`.  However, some things like like non-blocking IO on pipes isn't supported for anonymous pipes and changing the pipe buffer size is undocumented.

**Comment by milseman** [comment]
`ftruncate` as `FileDescriptor.resize`:  https://github.com/apple/swift-system/pull/82#pullrequestreview-964560242

**Comment by ffried** [comment]
@milseman Would `sendfile` be a candidate that could make it to this package? If so, I'd be happy to create a PR.

**Comment by milseman** [comment]
> @milseman Would `sendfile` (or maybe rather `copy_file_range`) be a candidate that could make it to this package? If so, I'd be happy to create a PR.

Yes, but I think that would be contingent on sockets, which got punted out (but hopefully we can readdress it soon)

**Comment by wibed** [comment]
where do block devices fit? should there be a seperate package depend on system provisioning block device support?
